pageserver/src/tenant/timeline.rs

```diff
@@ -4176,6 +4177,398 @@ impl Timeline {
+                    let copied = detach_ancestor::upload_rewritten_layer(
+                        end_lsn,
+                        &layer,
+                        &timeline,
+                        &timeline.cancel,
+                        &ctx,
+                    )
+                    .await?;
+                    Ok(copied)
+                });
+            }
+
+            while let Some(res) = tasks.join_next().await {
+                match res {
+                    Ok(Ok(Some(copied))) => {
+                        wrote_any = true;
+                        tracing::info!(layer=%copied, "rewrote and uploaded");
+                        new_layers.push(copied);
+                    }
+                    Ok(Ok(None)) => {}
+                    Ok(Err(e)) => return Err(e),
+                    Err(je) => return Err(Unexpected(je.into())),
+                }
+            }
+
+            // FIXME: the fsync should be mandatory, after both rewrites and copies in the case of
+            // failpoint usage
+            if wrote_any {
+                let timeline_dir = VirtualFile::open(
+                    &self
+                        .conf
+                        .timeline_path(&self.tenant_shard_id, &self.timeline_id),
+                )
+                .await
+                .fatal_err("VirtualFile::open for timeline dir fsync");
+                timeline_dir
+                    .sync_all()
+                    .await
+                    .fatal_err("VirtualFile::sync_all timeline dir");
+            }
+        }
+
+        // because we hold the layers, they will not be removed from remote storage.
+        let mut tasks = tokio::task::JoinSet::new();
+        let limiter = Arc::new(tokio::sync::Semaphore::new(options.copy_concurrency.get()));
+
+        for adopted in rest_of_historic {
+            let limiter = limiter.clone();
+            let timeline = self.clone();
+
+            tasks.spawn(
+                async move {
+                    let _permit = limiter.acquire().await;
+                    let owned = detach_ancestor::remote_copy(
+                        &adopted,
+                        &timeline,
+                        timeline.generation,
+                        &timeline.cancel,
+                    )
+                    .await?;
+                    tracing::info!(layer=%owned, "remote copied");
+                    Ok(owned)
+                }
+                .in_current_span(),
+            );
+        }
+
+        while let Some(res) = tasks.join_next().await {
+            match res {
+                Ok(Ok(owned)) => {
+                    new_layers.push(owned);
+                }
+                Ok(Err(failed)) => {
+                    return Err(failed);
+                }
+                Err(je) => return Err(Unexpected(je.into())),
+            }
+        }
+
+        // TODO: fsync directory again if we hardlinked something
```

Comment:
Seems like something that is trivially doable before merging this PR, flagging it. If it's not trivial, please add a TODO.

Reply:
No, I will not add the hardlinking in this PR.